src/ImageSharp/Formats/Qoi/QoiDecoderCore.cs

```diff
@@ -0,0 +1,291 @@
+
+        this.ProcessPixels(stream, pixels);
+
+        return image;
+    }
+
+    /// <inheritdoc />
+    public ImageInfo Identify(BufferedReadStream stream, CancellationToken cancellationToken)
+    {
+        ImageMetadata metadata = new();
+
+        this.ProcessHeader(stream);
+        PixelTypeInfo pixelType = new(8 * (int)this.header.Channels);
+        Size size = new((int)this.header.Width, (int)this.header.Height);
+
+        return new ImageInfo(pixelType, size, metadata);
+    }
+
+    /// <summary>
+    /// Processes the 14-byte header to validate the image and save the metadata
+    /// in <see cref="header"/>
+    /// </summary>
+    /// <param name="stream">The stream where the bytes are being read</param>
+    /// <exception cref="InvalidImageContentException">If the stream doesn't store a qoi image</exception>
+    private void ProcessHeader(BufferedReadStream stream)
+    {
+        Span<byte> magicBytes = stackalloc byte[4];
+        Span<byte> widthBytes = stackalloc byte[4];
+        Span<byte> heightBytes = stackalloc byte[4];
+
+        // Read magic bytes
+        int read = stream.Read(magicBytes);
+        if (read != 4 || !magicBytes.SequenceEqual(QoiConstants.Magic.ToArray()))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        // If it's a qoi image, read the rest of properties
+        read = stream.Read(widthBytes);
+        if (read != 4)
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        read = stream.Read(heightBytes);
+        if (read != 4)
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        // These numbers are in Big Endian so we have to reverse them to get the real number
+        uint width = BinaryPrimitives.ReadUInt32BigEndian(widthBytes),
+            height = BinaryPrimitives.ReadUInt32BigEndian(heightBytes);
+        if (width == 0 || height == 0)
+        {
+            throw new InvalidImageContentException(
+                $"The image has an invalid size: width = {width}, height = {height}");
+        }
+
+        int channels = stream.ReadByte();
+        if (channels is -1 or (not 3 and not 4))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        PixelTypeInfo pixelType = new(8 * channels);
+
+        int colorSpace = stream.ReadByte();
+        if (colorSpace is -1 or (not 0 and not 1))
+        {
+            ThrowInvalidImageContentException();
+        }
+
+        this.header = new QoiHeader(width, height, (QoiChannels)channels, (QoiColorSpace)colorSpace);
+    }
+
+    [DoesNotReturn]
+    private static void ThrowInvalidImageContentException()
+        => throw new InvalidImageContentException("The image is not a valid QOI image.");
+
+    private void ProcessPixels<TPixel>(BufferedReadStream stream, Buffer2D<TPixel> pixels)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        Rgba32[] previouslySeenPixels = new Rgba32[64];
```

Comment:
I though I already posted this:

```suggestion
        using IMemoryOwner<Rgba32> previouslySeenPixelsBuffer = this.memoryAllocator.Allocate<Rgba32>(64);
        Span<Rgba32> previouslySeenPixels = previouslySeenPixelsBuffer.GetSpan();
```

Reply:
I guess this won't be possible

The reference says that the array must be zero-initialized, and I though there was no problem since pixels will be overwritten but there's this error:

Due to `memoryAllocator`, the array isn't zero-initialized
![imagen](https://github.com/SixLabors/ImageSharp/assets/17792760/36e3a37a-818e-4fee-b928-8cf0e53c7342)

And if an decoded image starts with (0, 0, 0, 0) color (like `qoi_logo.qoi`), and it's encoded to qoi, then the first byte will be a 0b00000000, that means it's an QOI_OP_INDEX that requires the first position ([0]) of the previously seen pixels, but this won't be that color
![imagen](https://github.com/SixLabors/ImageSharp/assets/17792760/c82046b9-b564-409f-bf03-c5ee39a7dc71)

If the array is zero-initialized, then it's okay, it takes a (0, 0, 0, 0) color, however, using this implementation takes garbage of the previous decoding, so we must set everything to zero

I tried to use `previouslySeenPixelsBuffer.Dispose();` but it doesn't clear the array for the next decoding, is there any more efficient way to create the array? Or do I roll-back to `Rgba32[] previouslySeenPixels = new Rgba32[64];`?